conn/physic/units.go

```diff
@@ -733,3 +734,255 @@ func picoAsString(v int64) string {
+	digits string
+	// exponent is the left or right decimal shift. (powers of ten).
+	exp int
+	// neg it true if the number is negative.
+	neg bool
+}
+
+// Positive powers of 10 in the form such that powerOF10[index] = 10^index.
+var powerOf10 = [...]uint64{
+	1,
+	10,
+	100,
+	1000,
+	10000,
+	100000,
+	1000000,
+	10000000,
+	100000000,
+	1000000000,
+	10000000000,
+	100000000000,
+	1000000000000,
+	10000000000000,
+	100000000000000,
+	1000000000000000,
+	10000000000000000,
+	100000000000000000,
+	1000000000000000000,
+}
+
+// Maximum value for a int64.
+const maxInt64 = (1<<63 - 1)
+
+var maxUint64Str = "9223372036854775807"
+
+// Converts from decimal to int64, using the decimal.digits character values and
+// converting to a intermediate unit64.
+// Scale is combined with the decimal exponent to maximise the resolution and is
+// in powers of ten.
+func dtoi(d decimal, scale int) (int64, error) {
+	// Use uint till the last as it allows checks for overflows.
+	var u uint64
+	for i := 0; i < len(d.digits); i++ {
+		// Check that is is a digit.
+		if d.digits[i] >= '0' && d.digits[i] <= '9' {
+			// '0' = 0x30 '1' = 0x31 ...etc.
+			digit := d.digits[i] - '0'
+			// *10 is decimal shift left.
+			u *= 10
+			check := u + uint64(digit)
+			// Check should always be larger than u unless we have overflowed.
+			// Similarly if check > max it will overflow when converted to int64.
+			if check < u || check > maxInt64 {
+				if d.neg {
+					return (-maxInt64), &parseError{err: errors.New("overflows minimum is"), s: "-" + maxUint64Str}
+				}
+				return maxInt64, &parseError{err: errors.New("overflows maximum is"), s: maxUint64Str}
+			}
+			u = check
+		} else {
+			// Should not get here if used atod to generate the decimal.
+			return 0, &parseError{err: errors.New("not a number")}
+		}
+	}
+
+	// Get the total magnitude of the number.
+	// a^x * b^y = a*b^(x+y) since scale is of the order unity this becomes
+	// 1^x * b^y = b^(x+y).
+	// mag must be positive to use as index in to powerOf10 array.
+	mag := d.exp + scale
+	if mag < 0 {
+		mag *= -1
+	}
+	if mag > 18 {
+		return 0, errors.New("exponent exceeds int64")
+	}
+	// Divide is = 10^(-mag)
+	if d.exp+scale < 0 {
+		u = (u + powerOf10[mag]/2) / powerOf10[mag]
+	} else {
+		check := u * powerOf10[mag]
+		if check < u || check > maxInt64 {
+			if d.neg {
+				return -maxInt64, &parseError{err: errors.New("overflows minimum is"), s: "-" + maxUint64Str}
+			}
+			return maxInt64, &parseError{err: errors.New("overflows maximum is"), s: maxUint64Str}
+		}
+		u *= powerOf10[mag]
+	}
+
+	n := int64(u)
+	if d.neg {
+		n *= -1
+	}
+	return n, nil
+}
+
+// Converts a string to a decimal form. The return int is how many bytes of the
+// string are numeric. The string may contain +-0 prefixes and arbitrary
+// suffixes as trailing non number characters are ignored.
+// Significant digits are stored without leading or trailing zeros, rather an
+// exponet is used.
```

Comment:
exponent

Reply:
Oh the joys of dyslexia; the gift that keeps on giving.